src/core/eggmath.rkt

```diff
@@ -0,0 +1,213 @@
+#lang racket
+
+(require ffi/unsafe
+         ffi/unsafe/define)
+(require racket/runtime-path)
+
+(require "../debug.rkt" "../common.rkt")
+(require "../syntax/rules.rkt")
+(require "../syntax/herbie-to-egg-pattern.rkt")
+
+(require egg-herbie)
+
+(module+ test (require rackunit))
+
+(provide egraph-run egraph-add-exprs egraph-run-iter egraph-get-simplest egg-expr->expr egg-add-exn? make-ffi-rules egraph-get-cost egraph-get-size)
+
+
+(struct egraph-data (egraph-pointer egg->herbie-dict herbie->egg-dict))
+
+
+(define (egraph-get-size egraph-data)
+  (egraph_get_size (egraph-data-egraph-pointer egraph-data)))
+
+(define (egraph-get-cost egraph-data node-id)
+  (egraph_get_cost (egraph-data-egraph-pointer egraph-data) node-id))
+
+(define (egraph-get-simplest egraph-data node-id)
+  (egraph_get_simplest (egraph-data-egraph-pointer egraph-data) node-id))
+
+(define (make-ffi-rules rules)
+  (for/list [(rule rules)]
+      (make-FFIRule (symbol->string (rule-name rule))
+                    (herbie-pattern->rust-pattern (rule-input rule))
+                    (herbie-pattern->rust-pattern (rule-output rule)))))
+
+(define (egraph-run-iter egraph-data node-limit ffi-rules precompute?)
+  (egraph_run_iter (egraph-data-egraph-pointer egraph-data) node-limit ffi-rules precompute?))
+
+
+(define (egraph-run-rules egraph-data node-limit rules precompute?)
+  (define ffi-rules (make-ffi-rules rules))
+  (define old-cnt 0)
+  (for/and ([iter (in-naturals 0)])
+    (egraph_run_iter (egraph-data-egraph-pointer egraph-data) node-limit ffi-rules precompute?)
+    (define cnt (egraph_get_size (egraph-data-egraph-pointer egraph-data)))
+    (define is_stop (or (>= cnt node-limit) (<= cnt old-cnt)))
+    (set! old-cnt cnt)
+    (not is_stop)))
+
+
+;; calls the function on a new egraph, and cleans up
+(define (egraph-run egraph-function)
+  (define egraph (egraph-data (egraph_create) (make-hash) (make-hash)))
+  (define res (egraph-function egraph))
+  (egraph_destroy (egraph-data-egraph-pointer egraph))
+  res)
+
+(define (egg-expr->expr expr eg-data)
+  (define parsed (read (open-input-string expr)))
+  (egg-parsed->expr parsed (egraph-data-egg->herbie-dict eg-data)))
+
+(define (egg-parsed->expr parsed rename-dict)
+  (cond
+    [(list? parsed)
+     (cons
+      (first parsed)
+      (for/list ([expr (rest parsed)])
+        (egg-parsed->expr expr rename-dict)))]
+    [(and (number? parsed) (exact? parsed) (real? parsed))
+     parsed]
+    [else
+     (if (set-member? fpconstants parsed)
```

Comment:
`constant?`, and you could also merge the `if` statement upwards with the `cond`.